reference/matrix/ell_kernels.cpp

```diff
@@ -0,0 +1,142 @@
+
+
+#include "core/base/exception_helpers.hpp"
+#include "core/base/math.hpp"
+#include "core/matrix/dense.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+namespace ell {
+
+
+template <typename ValueType, typename IndexType>
+void spmv(std::shared_ptr<const ReferenceExecutor> exec,
+          const matrix::Ell<ValueType, IndexType> *a,
+          const matrix::Dense<ValueType> *b, matrix::Dense<ValueType> *c)
+{
+    auto max_nonzeros_per_row = a->get_max_nonzeros_per_row();
+    auto arows = a->get_num_rows();
+
+    for (size_type row = 0; row < arows; row++) {
+        for (size_type j = 0; j < c->get_num_cols(); j++) {
+            c->at(row, j) = zero<ValueType>();
+        }
+        for (size_type i = 0; i < max_nonzeros_per_row; i++) {
+            auto val = a->val_at(row, i);
+            auto col = a->col_at(row, i);
+            for (size_type j = 0; j < c->get_num_cols(); j++) {
+                c->at(row, j) += val * b->at(col, j);
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(GKO_DECLARE_ELL_SPMV_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void advanced_spmv(std::shared_ptr<const ReferenceExecutor> exec,
+                   const matrix::Dense<ValueType> *alpha,
+                   const matrix::Ell<ValueType, IndexType> *a,
+                   const matrix::Dense<ValueType> *b,
+                   const matrix::Dense<ValueType> *beta,
+                   matrix::Dense<ValueType> *c)
+{
+    auto max_nonzeros_per_row = a->get_max_nonzeros_per_row();
+    auto arows = a->get_num_rows();
+    auto valpha = alpha->at(0, 0);
+    auto vbeta = beta->at(0, 0);
+
+    for (size_type row = 0; row < arows; row++) {
+        for (size_type j = 0; j < c->get_num_cols(); j++) {
+            c->at(row, j) *= vbeta;
+        }
+        for (size_type i = 0; i < max_nonzeros_per_row; i++) {
+            auto val = a->val_at(row, i);
+            auto col = a->col_at(row, i);
+            for (size_type j = 0; j < c->get_num_cols(); j++) {
+                c->at(row, j) += valpha * val * b->at(col, j);
+            }
+        }
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+    GKO_DECLARE_ELL_ADVANCED_SPMV_KERNEL);
+
+
+template <typename ValueType, typename IndexType>
+void convert_to_dense(std::shared_ptr<const ReferenceExecutor> exec,
+                      matrix::Dense<ValueType> *result,
+                      const matrix::Ell<ValueType, IndexType> *source)
+{
+    auto num_rows = source->get_num_rows();
+    auto num_cols = source->get_num_cols();
+    auto max_nonzeros_per_row = source->get_max_nonzeros_per_row();
+
+    for (size_type row = 0; row < num_rows; row++) {
+        for (size_type col = 0; col < num_cols; col++) {
+            result->at(row, col) = zero<ValueType>();
+        }
+        for (size_type i = 0; i < max_nonzeros_per_row; i++) {
+            result->at(row, source->col_at(row, i)) += source->val_at(row, i);
```

Comment:
The `+=` reflects the situation when multiple values are assigned for the same location? I guess that is necessary to resolve the zero-padding, correct?

Reply:
Yes.
It can also resolve the two elements in the same row with the same index of ELL.
But I am not sure whether we need to consider this kind of situation.
I add some tests for Ell format with different padding.